src/mito2/src/schedule/scheduler.rs

```diff
@@ -0,0 +1,266 @@
+    /// State of scheduler
+    state: Arc<AtomicU8>,
+}
+
+impl LocalScheduler {
+    /// cap: flume bounded cap
+    /// rev_num: the number of bounded receiver
+    pub fn new(cap: usize, rev_num: usize) -> Self {
+        let (tx, rx) = flume::bounded(cap);
+        let token = CancellationToken::new();
+        let state = Arc::new(AtomicU8::new(STATE_RUNNING));
+
+        let mut handles = Vec::with_capacity(rev_num);
+
+        for _ in 0..rev_num {
+            let child = token.child_token().clone();
+            let receiver = rx.clone();
+            let state = Arc::clone(&state);
+            let handle = common_runtime::spawn_bg(async move {
+                while state.load(Ordering::Relaxed) == STATE_RUNNING {
+                    info!("Task scheduler loop.");
+                    tokio::select! {
+                        _ = child.cancelled() => {
+                            info!("Task scheduler cancelled.");
+                            // Wait all task finished
+                            if state.load(Ordering::Relaxed) == STATE_AWAIT_TERMINATION {
+                                while let Ok(req) = receiver.try_recv() {
+                                    req.await;
+                                }
+                                state.store(STATE_STOP, Ordering::Relaxed);
+                            }
+                            return;
+                        }
+                        req_opt = receiver.recv_async() =>{
+                            if let Ok(req) = req_opt {
+                                req.await;
+                            }
+                        }
+                    }
+                }
+            });
+            handles.push(handle);
+        }
+
+        Self {
+            sender: tx,
+            cancel_token: token,
+            handles: Mutex::new(handles),
+            state,
+        }
+    }
+
+    #[inline]
+    fn running(&self) -> bool {
+        self.state.load(Ordering::Relaxed) == STATE_RUNNING
+    }
+}
+
+#[async_trait::async_trait]
+impl Scheduler for LocalScheduler {
+    async fn schedule(&self, req: Job) -> Result<()> {
+        ensure!(
+            self.state.load(Ordering::Relaxed) == STATE_RUNNING,
+            InvalidSchedulerStateSnafu
+        );
+        self.sender
+            .send_async(req)
+            .await
+            .map_err(|_| InvalidFlumeSenderSnafu {}.build())
+    }
+
+    async fn stop(&self, await_termination: bool) -> Result<()> {
+        ensure!(
+            self.state.load(Ordering::Relaxed) == STATE_RUNNING,
+            InvalidSchedulerStateSnafu
+        );
+        let state = if await_termination {
+            STATE_AWAIT_TERMINATION
+        } else {
+            STATE_STOP
+        };
+        self.state.store(state, Ordering::Relaxed);
+        self.cancel_token.cancel();
+
+        for handle in self.handles.lock().await.drain(..) {
+            handle.await.context(StopSchedulerSnafu)?;
+        }
```

Comment:
nit: In some cases we'd prefer [`futures::future::try_join_all`](https://docs.rs/futures/latest/futures/future/fn.try_join_all.html) instead of iterating and waiting for each individual future since in the former futures are driven all together while in the latter futures are driven one after another:

```suggestion
        futures::future::join_all(self.handles.lock().await.drain(..))
            .await
            .into_iter()
            .collect::<std::result::Result<Vec<_>, _>>()
            .context(StopSchedulerSnafu)?;
```